src/encoding/bencode.rs

```diff
@@ -1,136 +1,125 @@
 #[derive(Debug)]
 pub enum Error {
-    FromUtf8,
     NonStandardChar(char)
 }
 
-pub trait Literal {
-    fn stringify(&self) -> Result<String, Error>;
-}
-
-// non primative bencode types
-pub type ListEntry = Box<dyn Literal>;
-pub type DictEntry = (ByteStr, Box<dyn Literal>);
 
-/*
-**  Bencode data structures
-**  (https://wiki.theory.org/index.php/BitTorrentSpecification#Bencoding)
-*/
+pub trait Encodable {
+    fn encode(&self) -> Result<String, Error>;
+}
 
 /*
 **  Byte strings are encoded as follows: <string length encoded in base ten ASCII>:<string data>
 **  Note that there is no constant beginning delimiter, and no ending delimiter.
 */
+impl Encodable for &str {
+    fn encode(&self) -> Result<String, Error> {
+        let mut bstring: Vec<char> = Vec::new();
+        for c in self.chars() {
+            if c as u32 > 127 {
+                return Err(Error::NonStandardChar(c))
+            }
 
-pub struct ByteStr {
-    content: Vec<u8>
-}
-
-impl Literal for ByteStr {
-    // create a String representation from ASCII sequence
-    fn stringify(&self) -> Result<String, Error> {
-        let arr = &self.content;
-        let ascii_result = String::from_utf8(arr.to_vec());
-        match ascii_result {
-            Ok(trailer) => Ok(format!("{}:{}", arr.len(), trailer)),
-            Err(_) => Err(Error::FromUtf8)
+            bstring.push(c);
         }
-    }
-}
 
-impl ByteStr {
-    // create a ByteStr from a string slice
-    pub fn try_from_str(ascii_128_str: &str) -> Result<Self, Error> {
-        let mut content: Vec<u8> = vec![];
 
-        // allocate chars to ASCII decimal vec
-        for c in ascii_128_str.chars() {
-            let decimal = c as usize;
-            // char incompatible with bencode
-            if decimal > 127 {
-                return Err(Error::NonStandardChar(c))
-            }
-            content.push(c as u8);
-        }
+        Ok(
+            format!("{}:{}", bstring.len(), bstring.iter().collect::<String>())
+        )
+    }
+}
 
-        Ok(ByteStr { content })
+impl Encodable for String {
+    fn encode(&self) -> Result<String, Error> {
+        Ok(Encodable::encode(&&self[..])?)
     }
 }
 
+
 /*
 **  Integers are encoded as follows: i<integer encoded in base ten ASCII>e
 **  The initial i and trailing e are beginning and ending delimiters.
 */
-
-pub struct Int {
-    content: isize
-}
-
-impl Literal for Int {
-
-    fn stringify(&self) -> Result<String, Error> {
-        Ok(format!("i{}e", self.content))
-    }
+macro_rules! encode_int {
+    ($int: ty) => {
```

Comment:
- Is `$int` just a placeholder for a generic type?
- What does `ty` mean, Is this just a type?

Reply:
`$int` is the variable being passed into the macro. 
It is a type as denoted by `ty`.

https://doc.rust-lang.org/rust-by-example/macros.html

Reply:
🚀🚀🚀